packages/gatsby/src/utils/loading-indicator.ts

```diff
@@ -0,0 +1,52 @@
+import { Express } from "express"
+import { writeModule } from "./gatsby-webpack-virtual-modules"
+
+// set value to undefined first, because env vars needed to determine if indicator
+// should ever be enabled by default might not be set yet - we set it to "initial"
+// first time we write out module if loading indicator is allowed
+// "initial" means that browser will decide if it should show it
+// for now we do disable it by default when running in cypress
+// to not cause problems for users when they iterate on their E2E tests
+// this check could be expanded in the future to add support for more scenarios
+// like that.
+let indicatorEnabled: "initial" | true | false | undefined = undefined
+
+export function writeVirtualLoadingIndicatorModule(): void {
+  if (indicatorEnabled === undefined) {
+    indicatorEnabled =
+      process.env.GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND &&
+      process.env.GATSBY_QUERY_ON_DEMAND_LOADING_INDICATOR === `true`
+        ? `initial`
+        : false
+  }
+
+  writeModule(
+    `$virtual/loading-indicator.js`,
+    `
+    export function isLoadingIndicatorEnabled() {
+    return ${
+      indicatorEnabled === `initial`
+        ? `\`Cypress\` in window
+          ? false
+          : true`
+        : JSON.stringify(indicatorEnabled)
+    }
+  }`
+  )
+}
+
+export function routeLoadingIndicatorRequests(app: Express): void {
+  app.get(`/___loading-indicator/:method?`, (req, res) => {
+    if (req.params.method === `enable` && indicatorEnabled !== true) {
+      indicatorEnabled = true
+      writeVirtualLoadingIndicatorModule()
+    } else if (req.params.method === `disable` && indicatorEnabled !== false) {
+      indicatorEnabled = false
```

Comment:
Does it make sense to disable it for all develop sessions (i.e. save to cache)? Otherwise, users will have to disable it on every `gatsby develop` run?

Reply:
For permanent solution I would want to be able to configure this in `gatsby-config` - I even had this before, but it was setting precedence that we would have to stick with and it didn't feel right to do it in haste.

https://github.com/gatsbyjs/gatsby/pull/28562/commits/0a180e5f45d9dd2ef6f5519ed62945654ca25440 removed permanent setting

I did put note about this in description:
> ~[x] permenently (gatsby-config)~ I backed out of it, because it sets precedent for configuration in flags which we might regret - it doesn't mean something like won't be added in the future - just that it needs more thinking and discussions

I don't like using this method to save permanently, because it's only thing that act this way and it's not clear to users how to re-enable it later.

Disabling for current session is mostly here to address potential issues of adding "foreign UI elements" to user sites - like iterating on e2e tests (i.e. cypress) in develop or running some other forms of automated audits with browser extensions etc.